Staff of Shrinking for the wizard

PKPenguin321 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request

Adds a new staff for the wizard that shoots shrink rays. Also a corresponding wand that comes with the wand belt. Shrinking is a mechanic already implemented by abductors, but it's not often used because it doesn't fit their kit super well. That's a huge shame because shrinking stuff/people is really funny. And you know where funny stuff fits well? The wizard kit.

OH YEAH and being shrunken now gives you the squash component so you can be squashed as though you were a roach, though this only deals 10 damage instead of gibbing you

tiny staff
![68747470733a2f2f696462322e616666656374656461726330372e636f2e756b2f696d616765732f32353836323339392f38333131352f31333237383435323531373936343834343132322e706e67](https://github.com/tgstation/tgstation/assets/8345184/af777602-036f-4171-ad43-5dafa7b29414)
tiny wand
![68747470733a2f2f696462322e616666656374656461726330372e636f2e756b2f696d616765732f32353836323339392f38333131352f393439333932353034373033303238393530372e706e67](https://github.com/tgstation/tgstation/assets/8345184/81241789-404f-4d8e-8473-2ec74b171f55)
exhausted wand turns back to a big wand sprite :)
![68747470733a2f2f696462322e616666656374656461726330372e636f2e756b2f696d616765732f32353836323339392f38333131352f323039323438333731323537303338313239362e706e67](https://github.com/tgstation/tgstation/assets/8345184/c2073952-c16b-4113-9ca9-527cfdbcfd16)


## Why It's Good For The Game

Shrinking stuff is funny, plus it gives the wizard something new to do besides polymorphing everyone or turning everybody to stone or ei nathing people.

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->

:cl:
add: New funny wizard staff/wand that shrinks stuff.
add: Being shrunken now leaves you vulnerable to being crushed to death.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

PKPenguin321 committed:
shrink magic

Sadboysuss commented:
this could be treated as a buff to abductors since they have a shrink ray gun

PKPenguin321 commented:
yeh but theyre not supposed to murder folks anyway so

ElGitificador commented:
any chance you make the shrink permanent? it d be hilarous to have a miniaturized crew after a wizard attack. pretty ss13 if you ask me

improvedname commented:
make the wizard shrink perma for shit n giggles while abductor shrink is timed

SomeoneYouProbablyKnow commented:
Permanent shrinking would be very funny as long as you don't put that on the wand belt, every second wizard gets it wouldnt want to make it mandatory

Sadboysuss commented:
> Yeah I could make it permanent and drop the wand charges down a bit, sure Also would it be broken to make shrunken items a smaller w_class

yes

PKPenguin321 commented:
Should I do that anyway

SomeoneYouProbablyKnow commented:
As long as smaller backpacks carry less stuff so it evens out for shrunk people? but smells of feature creep ngl

jughu commented:
make it half damage by each weight level it shrunk

MrMelbert commented on code/datums/components/shrink.dm:
```diff
@@ -1,6 +1,7 @@
 /datum/component/shrink
 	var/olddens
 	var/oldopac
+	var/newsquash = null
```
```suggestion
	/// Tracks the squashable component we apply when we make the small mob squashable
	var/atum/component/squashable/newsquash
```

MrMelbert commented on code/datums/components/shrink.dm:
```diff
@@ -25,6 +26,13 @@
 			if(ishuman(C))
 				var/mob/living/carbon/human/H = C
 				H.physiology.damage_resistance -= 100//carbons take double damage while shrunk
+		if(!L.GetComponent(/datum/component/squashable))
+			newsquash = L.AddComponent( \
+				/datum/component/squashable, \
+				squash_chance = 75, \
+				squash_damage = 10, \
+				squash_flags = SQUASHED_ALWAYS_IF_DEAD|SQUASHED_DONT_SQUASH_IN_CONTENTS, \
+			)
```
Using `GetComponent` is generally bad, and components should handle what happens if a datum already has the passed component:
```suggestion
		newsquash = L.AddComponent( \
			/datum/component/squashable, \
			squash_chance = 75, \
			squash_damage = 10, \
			squash_flags = SQUASHED_ALWAYS_IF_DEAD|SQUASHED_DONT_SQUASH_IN_CONTENTS, \
		)
```
Note, `/datum/component/squashable` is `dupe_mode = COMPONENT_DUPE_HIGHLANDER`. 
Meaning it will replace existing ones. Which is probably fine? 
If you want to avoid doing that, you would need to change its dupe mode.

PKPenguin321 replied:
Yeah I wanted to avoid doing, I guess I could replace the dupe mode but I just wasn't sure if anything else relied on it

PKPenguin321 replied:
Actually wouldn't this introduce a potential issue where, if the mob was previously squishable like a roach, they wouldn't be after unshrinking?

MrMelbert replied:
Eh that's a good point, I forgot that `AddComponent` returns existing instances in the event of a dupe mode conflict. Fairly certain this is still solvable with dupe modes in some way but I'm not sure, so leave it as it I suppose

MrMelbert commented on code/datums/components/shrink.dm:
```diff
@@ -36,6 +44,8 @@
 	speech_args[SPEECH_SPANS] |= SPAN_SMALL_VOICE
 
 /datum/component/shrink/Destroy()
+	if(newsquash)
+		qdel(newsquash)
```
```suggestion
	QDEL_NULLl(newsquash)
```

MrMelbert commented on code/modules/projectiles/guns/magic/staff.dm:
```diff
@@ -319,3 +320,17 @@
+	recharge_rate = 5
+	no_den_usage = 1
+	school = SCHOOL_TRANSMUTATION
+	slot_flags = null //too small to wear on your back
```
```suggestion
	slot_flags = NONE //too small to wear on your back
```

MrMelbert commented on code/modules/projectiles/guns/magic/wand.dm:
```diff
@@ -256,3 +256,25 @@
+	to_chat(user, span_notice("The world grows large..."))
+	charges--
+	user.AddComponent(/datum/component/shrink, 900)
+	. = ..()
```
```suggestion
	return ..()
```

MrMelbert commented on code/modules/projectiles/guns/magic/wand.dm:
```diff
@@ -256,3 +256,25 @@
+/obj/item/gun/magic/wand/shrink/zap_self(mob/living/user)
+	to_chat(user, span_notice("The world grows large..."))
+	charges--
+	user.AddComponent(/datum/component/shrink, 900)
```
```suggestion
	user.AddComponent(/datum/component/shrink, 90 SECONDS)
```

MrMelbert commented on code/modules/projectiles/projectile/magic.dm:
```diff
@@ -580,3 +580,10 @@
 	antimagic_charge_cost = 0 // since the cards gets spammed like a shotgun
+
+/obj/projectile/beam/shrink/magic
+	shrink_time = 900
```
```suggestion
	shrink_time = 90 SECONDS
```

MrMelbert commented on code/modules/projectiles/projectile/magic.dm:
```diff
@@ -580,3 +580,10 @@
+	shrink_time = 900
+
+/obj/projectile/beam/shrink/magic/on_hit(atom/target, blocked = 0, pierce_hit)
+	shrink_time = rand(600, 900)
+	. = ..()
```
```suggestion
	shrink_time = rand(60 SECONDS, 90 SECONDS)
	return ..()
```

MrMelbert commented on code/modules/projectiles/projectile/magic.dm:
```diff
@@ -580,3 +580,10 @@
 	damage = 2
 	antimagic_charge_cost = 0 // since the cards gets spammed like a shotgun
+
+/obj/projectile/beam/shrink/magic
```
Because this is not subtyped off of `/obj/projectile/magic`, this will not be blocked by antimagic. Which is not ideal?

PKPenguin321 replied:
Hmmm. Yeah I was wondering if that would matter. Might just swap the whole thing over to the magic subtype and abductors can just have deep lore that they're secretly using magic or something

MrMelbert replied:
That would work for me

tralezab commented:
wouldn't be a bad wand of chaos bolt

PKPenguin321 commented:
Yeah it's on the chaos staff too

GuillaumePrata commented:
Could make them all permanent, but being blessed by the chaplain, eating a holy melon, or whatever else makes sense, turn you back to normal.

Now it is on the victims to choose if they care enough to go back, because shrunk station staff does sound a lot funnier for the round.

PKPenguin321 committed:
suggested changes, small forever

PKPenguin321 commented:
addressed requested changes and added sprites to PR body. shrinking from the staff only (not the wand or abductor gun) is now permanent

known issue:
- if you hit someone with a shrink ray when theyre already shrunk it removes the squash component? and also the trait that makes you able to pass through them? not sure what the deal is, some component nonsense, help appreciated

possible maybe changes to shrinking:
- the slowdown is pretty brutal, permanent shrinking seems a bit miserable with how slow you go. maybe that could be temporary on hit but the rest of the shrink can be normal idk
- the remove by chaplain idea is cool and shouldnt be super hard
- i think it would be cool if shrunk structures had their armor and HP severely reduced
- i think it would be cool if small items had their damage and weight class severely reduced but yeah maybe thats a bit broken idk
- it would be nice if small people couldnt squish other small people, not sure how that works. do mobs have a size tracker on them nowadays or anything convenient like that?

carlarctg commented:
> Yeah I could make it permanent and drop the wand charges down a bit, sure
> Also would it be broken to make shrunken items a smaller w_class

yes but so is wizard as a whole. this sounds incredibly fun and 'interacts with the sandbox' which is what all the mantainers love

MrDoomBringer commented:
make it lower w_class and if it really sucks we can just undo it

Ghommie commented on code/modules/projectiles/guns/magic/staff.dm:
```diff
@@ -319,3 +320,17 @@
+	inhand_icon_state = "staff"
+	max_charges = 10 // slightly more/faster charges since this will be used on walls and such
+	recharge_rate = 5
+	no_den_usage = 1
```
```suggestion
	no_den_usage = TRUE
```

Ghommie commented on code/modules/projectiles/guns/magic/wand.dm:
```diff
@@ -256,3 +256,25 @@
+	base_icon_state = "shrinkwand"
+	fire_sound = 'sound/magic/staff_shrink.ogg'
+	max_charges = 10 //10, 5, 5, 4
+	no_den_usage = 1
```
```suggestion
	no_den_usage = TRUE
```

PKPenguin321 committed:
1 to TRUE

ATH1909 commented:
> > Yeah I could make it permanent and drop the wand charges down a bit, sure
> > Also would it be broken to make shrunken items a smaller w_class
> 
> yes but so is wizard as a whole. this sounds incredibly fun and 'interacts with the sandbox' which is what all the mantainers love

wizard stuff is intentionally OP, but balancing levers (that aren't trivial to circumvent), such as item size, are still nice to have.

PKPenguin321 commented:
alright cool
